lib/id3c/cli/command/etl/fhir.py

```diff
@@ -0,0 +1,740 @@
+    find_location,
+    find_sample,
+    upsert_individual,
+    upsert_encounter,
+    upsert_encounter_location,
+    upsert_location,
+    upsert_sample,
+    upsert_presence_absence,
+
+    SampleNotFoundError,
+)
+
+
+LOG = logging.getLogger(__name__)
+
+
+# The revision number and etl name are stored in the processing_log of each
+# FHIR record when the FHIR document is successfully processed by this ETL
+# routine. The routine finds new-to-it records to process by looking for FHIR
+# documents lacking this etl revision number and etl name in their log.
+# If a change to the ETL routine necessitates re-processing all FHIR documents,
+# this revision number should be incremented.
+# The etl name has been added to allow multiple etls to process the same
+# receiving table
+REVISION = 1
+ETL_NAME = 'fhir'
+INTERNAL_SYSTEM = 'https://seattleflu.org'
+LOCATION_RELATION_SYSTEM = 'http://terminology.hl7.org/CodeSystem/v3-RoleCode'
+SNOMED_SYSTEM = 'http://snomed.info/sct'
+EXPECTED_COLLECTION_IDENTIFIER_SETS = [
+    'collections-household-observation',
+    'collections-household-intervention',
+    'collections-swab&send',
+    'collections-kiosks',
+    'collections-self-test',
+]
+
+@etl.command("fhir", help = __doc__)
+
+@with_database_session
+def etl_fhir(*, db: DatabaseSession):
+    LOG.debug(f"Starting the FHIR ETL routine, revision {REVISION}")
+
+    # Fetch and iterate over FHIR documents that aren't processed
+    #
+    # Use a server-side cursor by providing a name.  This ensures we limit how
+    # much data we fetch at once, to limit local process size.
+    #
+    # Rows we fetch are locked for update so that two instances of this
+    # command don't try to process the same FHIR documents.
+    LOG.debug("Fetching unprocessed FHIR documents")
+
+    fhir_documents = db.cursor("fhir")
+    fhir_documents.execute("""
+        select fhir_id as id, document
+          from receiving.fhir
+         where not processing_log @> %s
+         order by id
+           for update
+        """, (Json([{ "etl": ETL_NAME, "revision": REVISION }]),))
+
+    for record in fhir_documents:
+        with db.savepoint(f"FHIR document {record.id}"):
+            LOG.info(f"Processing FHIR document {record.id}")
+
+            assert_bundle_collection(record.document)
+            bundle      = Bundle(record.document)
+            resources   = extract_resources(bundle)
+
+            try:
+                assert_required_resource_types_present(resources)
+            except AssertionError:
+                LOG.info("FHIR document doesn't meet minimum requirements for processing.")
+                mark_skipped(db, record.id)
+                continue
+
+            # Loop over every Resource the Bundle entry, processing what is
+            # needed along the way.
+            try:
+                process_bundle_entries(db, bundle)
+
+            except AssertionError:
+                LOG.warning("Insufficient Encounter information.")
```

Comment:
Catching `AssertionError` here will catch more exceptions than the one you're anticipating from `process_encounter_bundle_entry()` (for example, there's an assert in `upsert_encounter`). I think a custom exception class is warranted here to more precisely target the catch. Ditto on the similar catch above, which is admittedly more tightly scoped. We want other assertions to still cause the whole process to bomb out and fail so we can investigate and intervene.

Reply:
I think this is a blocker as it can cause silent-ish failure.